src/source/agc.rs

```diff
@@ -77,55 +144,41 @@ where
     #[inline]
     fn next(&mut self) -> Option<I::Item> {
         self.input.next().map(|value| {
+            // Convert the sample to its absolute float value for level calculations
             let sample_value = value.to_f32().abs();
 
-            // Update peak level with adaptive attack coefficient
-            let attack_coeff = if sample_value > self.peak_level {
-                self.attack_coeff.min(0.1) // Faster response to sudden increases
-            } else {
-                self.attack_coeff
-            };
-            self.peak_level = attack_coeff * self.peak_level + (1.0 - attack_coeff) * sample_value;
-
-            // Update RMS level using a sliding window
-            self.rms_level -= self.rms_window[self.rms_index] / self.rms_window.len() as f32;
-            self.rms_window[self.rms_index] = sample_value * sample_value;
-            self.rms_level += self.rms_window[self.rms_index] / self.rms_window.len() as f32;
-            self.rms_index = (self.rms_index + 1) % self.rms_window.len();
+            // Dynamically adjust peak level using an adaptive attack coefficient
+            self.update_peak_level(sample_value);
 
-            let rms = self.rms_level.sqrt();
+            // Calculate the current RMS (Root Mean Square) level using a sliding window approach
+            let rms = self.update_rms(sample_value);
 
-            // Calculate gain adjustments based on peak and RMS levels
-            let peak_gain = if self.peak_level > 0.0 {
-                self.target_level / self.peak_level
-            } else {
-                1.0
-            };
+            // Determine the gain adjustment needed based on the current peak level
+            let peak_gain = self.calculate_peak_gain();
 
+            // Compute the gain adjustment required to reach the target level based on RMS
             let rms_gain = if rms > 0.0 {
                 self.target_level / rms
             } else {
-                1.0
+                1.0 // Default to unity gain if RMS is zero to avoid division by zero
             };
 
-            // Choose the more conservative gain adjustment
+            // Select the lower of peak and RMS gains to ensure conservative adjustment
             let desired_gain = peak_gain.min(rms_gain);
 
-            // Set target gain to the middle of the allowable range
-            let target_gain = 1.0; // Midpoint between 0.1 and 3.0
```

Comment:
Removed `target_gain` since it seems redundant. However, it's a potential area to monitor for any issues that may arise that I couldn't identify/find.